### PR TITLE
Upgrade Fantomas and add CheckFormat to build pipeline

### PR DESCRIPTION
**Pull request by yisusalanpng**
@nojaf Could you give this a look please? 
Also does .editorconfig needs changing?

### CODE REVIEW

**Comment by nojaf** [comment]
Hello @sergey-tihon, @yisusalanpng is an intern at the Open-Source team at G-Research.
He will be helping me out with some Fantomas improvements and as a learning exercise, I've asked him to upgrade Fantomas to the latest version. I hope you don't mind this PR where we update Fantomas and enforce formatting during the FAKE build.

@yisusalanpng you will need to check the [Changelog file](https://github.com/fsprojects/fantomas/blob/master/CHANGELOG.md#500-alpha-002---2022-04-04) to deduce if you need to change any `.editorconfig` settings. Spoiler: some settings were renamed so you definitely need to address a couple of things.

**Comment by yisusalanpng** [comment]
All set! Looks like the same config got renamed twice, so only that one setting needed changing. 
Thanks a lot!

**Comment by sergey-tihon** [comment]
> Spoiler: some settings were renamed so you definitely need to address a couple of things.

Is there a "migration" story for setting? How Fantomas user should keep editorconfig up-to-date? Read release notes?

**Comment by nojaf** [comment]
I'll add some notes on how to upgrade from v4 to v5 (stable).

This is a bit of a unicorn situation where a project is using a feature that was still somewhat in development. I'm very grateful you are doing this, as it can provide some insights into how this new feature faires in the wild.

Our changelog is the source of truth to know what changed between releases.

File: .editorconfig [hunk 1]
```diff
@@ -21,4 +21,4 @@ fsharp_max_infix_operator_expression=70
 fsharp_space_before_parameter=false
 fsharp_space_before_lowercase_invocation=false
 fsharp_multiline_block_brackets_on_same_column=true
-fsharp_ragnarok=true
+fsharp_experimental_stroustrup_style=true
```
Comment (nojaf) [inline .editorconfig:24]: @sergey-tihon should @yisusalanpng apply the same settings for `*.fsx` as well?

Comment (sergey-tihon) [reply .editorconfig:24]: Good idea, I do not mind, thanks.

File: src/SwaggerProvider.DesignTime/Provider.OpenApiClient.fs [hunk 2]
```diff
@@ -44,47 +58,70 @@ type public OpenApiClientTypeProvider(cfg : TypeProviderConfig) as this =
...
+                with _ ->
+                    Cache.providedTypes.Remove(cacheKey) |> ignore
+
+                    Cache
+                        .providedTypes
+                        .GetOrAdd(
+                            cacheKey,
+                            addCache
+                        )
+                        .Value
```
Comment (sergey-tihon) [inline src/SwaggerProvider.DesignTime/Provider.OpenApiClient.fs:122]: @nojaf @yisusalanpng  is it 'ideomatic' formatting? `GetOrAdd` call looks strange for me
```
                    Cache
                         .providedTypes
                         .GetOrAdd(
                             cacheKey,
                             addCache
                         )
                         .Value
```

File: src/SwaggerProvider.DesignTime/v2/Parser/SchemaParserExceptions.fs [hunk 3]
```diff
@@ -6,30 +6,26 @@ type SwaggerSchemaParseException(message) =
...
-    inherit SwaggerSchemaParseException(
-        sprintf "Value `%s` is not allowed for field `%s`(See %s for more details).\nObject:%A"
-            value field specLink obj)
+type UnknownFieldValueException<'T>(obj: 'T, value: string, field: string, specLink: string) =
+    inherit SwaggerSchemaParseException(sprintf
+                                            "Value `%s` is not allowed for field `%s`(See %s for more details).\nObject:%A"
+                                            value
+                                            field
+                                            specLink
+                                            obj)
```
Comment (sergey-tihon) [inline src/SwaggerProvider.DesignTime/v2/Parser/SchemaParserExceptions.fs:19]: @nojaf @yisusalanpng here I also a little bit surprised
```fsharp
    inherit SwaggerSchemaParseException(sprintf
                                             "Value `%s` is not allowed for field `%s`(See %s for more details).\nObject:%A"
                                             value
                                             field
                                             specLink
                                             obj)
```

why Fantoms does not insert line break before `sprintf`?

File: src/SwaggerProvider.DesignTime/Provider.OpenApiClient.fs (hunk 2 above)
Comment (nojaf) [reply src/SwaggerProvider.DesignTime/Provider.OpenApiClient.fs:122]: There are no real rules about this in the style guide.
I would encourage you to raise this concern or proposal to https://github.com/fsharp/fslang-design#style-guide.
I'm in favour of something better, but the community should come up with some guidance.

Comment (sergey-tihon) [reply src/SwaggerProvider.DesignTime/Provider.OpenApiClient.fs:122]: Ok, done https://github.com/fsharp/fslang-design/issues/688

File: .editorconfig (hunk 1 above)
Comment (yisusalanpng) [reply .editorconfig:24]: sure thing! also please let me know if I should add the 'tests' folder to the scope as well :)

File: src/SwaggerProvider.DesignTime/Provider.OpenApiClient.fs (hunk 2 above)
Comment (nojaf) [reply src/SwaggerProvider.DesignTime/Provider.OpenApiClient.fs:122]: Thank you.

**Comment by nojaf** [comment]
@sergey-tihon is this good to go?
The reported bug and style guide questions won't be solved overnight (and are most likely present in the current version of Fantomas anyway).

File: .editorconfig (hunk 1 above)
Comment (sergey-tihon) [reply .editorconfig:24]: @yisusalanpng yes, please, let's format all F# source files in this repo

**Comment by sergey-tihon** [comment]
Thank you @yisusalanpng & @nojaf  for this PR

Would be nice if you can try to format https://github.com/fsprojects/FSharp.TypeProviders.SDK as well.
I believe that Fantomas still fails to parse https://github.com/fsprojects/FSharp.TypeProviders.SDK/blob/master/src/ProvidedTypes.fs

File: .config/dotnet-tools.json [hunk 4]
```diff
@@ -20,11 +20,11 @@
         "dotnet-serve"
       ]
     },
-    "fantomas-tool": {
-      "version": "5.0.0-alpha-002",
+    "fantomas": {
+      "version": "5.0.0-alpha-010",
```
Comment (nojaf) [inline .config/dotnet-tools.json:24]: Ah shoot, this is already merged. @yisusalanpng I released alpha 11 on Friday. Would you mind making another PR?

**Comment by nojaf** [comment]
Great, we can take a look at that as well.

**Pull request merged**